### PR TITLE
Only create new ctypes for unqualified struct and enum types

### PR DESCRIPTION
**Pull request by yebblies**
The current code creates a new type from scratch only if `t->mutableOf()->ctype` is not set, which will only happen for `T` and `shared(T)`.  Otherwise it copies the ctype and adds the missing qualifiers.

Instead, create a new ctype only if the type is unqualified, and otherwise generate the unqualified ctype first and copy it to add qualifiers.

### CODE REVIEW

**Comment by yebblies** [comment]
All green.

File: src/toctype.c [hunk 1]
```diff
@@ -182,15 +182,21 @@ class ToCtypeVisitor : public Visitor
     void visit(TypeEnum *t)
     {
         //printf("TypeEnum::toCtype() '%s'\n", t->sym->toChars());
-        Type *tm = t->mutableOf();
-        if (tm->ctype && tybasic(tm->ctype->Tty) == TYenum)
+        if (t->sym->memtype->toBasetype()->ty == Tint32)
```
Comment (ibuclaw) [inline src/toctype.c]: Out of curiosity, is this a dmd limitation?  You should be able to make all integral types (but not bool) as enums.

File: src/toctype.c [hunk 2]
```diff
@@ -174,6 +137,43 @@ class ToCtypeVisitor : public Visitor
                     symbol_struct_addField(t->ctype->Ttag, v->ident->toChars(), Type_toCtype(v->type), v->offset);
                 }
             }
+            return;
+        }
+
+        // Copy the ctype from the mutable version and add mods
+        type *mctype = Type_toCtype(t->mutableOf()->unSharedOf());
```
Comment (ibuclaw) [inline src/toctype.c]: Maybe use castMod(0)?

**Comment by ibuclaw** [comment]
FYI gdc does this for all types, not just struct/enums.  Not sure if applicable for dmd, but it does help improve debug code generation.

File: src/toctype.c (hunk 1 above)
Comment (yebblies) [inline src/toctype.c]: My guess would be a codeview limitation.

File: src/toctype.c (hunk 2 above)
Comment (yebblies) [inline src/toctype.c]: Does that do the same thing?  I wasn't sure.

**Comment by yebblies** [comment]
Are you saying GDC provides full type information to the backend?  I'm not sure generating debug information in the dmd backend is the right approach, given the difficulties of working with it.

File: src/toctype.c (hunk 2 above)
Comment (ibuclaw) [inline src/toctype.c]: Yes, it does the necessary toXXXOf conversions to satisfy 'mod'

**Comment by ibuclaw** [comment]
> Are you saying GDC provides full type information to the backend?

Yes, how else is the type-strict optimiser to work out that ``const(T)`` and ``shared(T)`` derive from the same type, not distinct types that bear no equality?  (This applies for all types that are represented as structs - D arrays, AAs, delegates).

> I'm not sure generating debug information in the dmd backend is the right approach, given the difficulties of working with it.

Enough people use DMD that it's a burden we must live with.

**Comment by ibuclaw** [comment]
@yebblies - This would be the equivalent of doing it in ``toCtype`` for you. https://github.com/ibuclaw/GDC/commit/e2557ad6608db605b577703b9c65e5eb5e887571#diff-4b1c9aa130668959e9a51517f9040224R489

File: src/toctype.c (hunk 2 above)
Comment (yebblies) [inline src/toctype.c]: Done

**Comment by yebblies** [comment]
> Enough people use DMD that it's a burden we must live with.

I dunno, it's probably possible to generate it in the glue layer, at least for type info.

**Comment by yebblies** [comment]
It's passing now, @ibuclaw could you pull it?

**Comment by ibuclaw** [comment]
Remind me in a few hours to review.  :)

**Comment by yebblies** [comment]
> Remind me in a few hours to review. :)

!

**Comment by ibuclaw** [comment]
Auto-merge toggled on

**Comment by yebblies** [comment]
Thanks!

**Pull request merged**